### PR TITLE
RDSK-9311: (base/board) Guard against nil responses in rdk client/servers

### PR DESCRIPTION
**Pull request by martha-johnston**
went through board and base server code to see what had the possibility to be returned as nil, and added checks to return errors in that case. also added tests for each nil failure.

added necessary tests to client test file to ensure that the server changes resulted in the same returned error, so the checks didn't need to be added in the client files as well. 

still need to do this for the other components, but want to break up the work into smaller PRs (https://github.com/viamrobotics/rdk/pull/4620 & https://github.com/viamrobotics/rdk/pull/4621), as board required quite a few changes

### CODE REVIEW

**Review by randhid — COMMENTED** [review]
Nice testing and cleanup, good choice to break the work up into three! Can you also add any manual tests performed with this code?

File: components/base/server.go [hunk 1]
```diff
@@ -162,6 +168,9 @@ func (s *serviceServer) GetGeometries(ctx context.Context, req *commonpb.GetGeom
 	if err != nil {
 		return nil, err
 	}
+	if geometries == nil {
+		return nil, ErrGeometriesNil(req.GetName())
+	}
 	return &commonpb.GetGeometriesResponse{Geometries: spatialmath.NewGeometriesToProto(geometries)}, nil
```
Comment (martha-johnston) [inline components/base/server.go:174]: @cheukt `NewGeometriesToProto` doesn't return an error, so I can add an error return, but that just requires me to change everywhere else it's used. it's just a couple components, but feels more breaking than just adding a nil check before we even get there. thoughts on leaving the geometries implementation at the client/server level?

**Comment by martha-johnston** [comment]
@JohnN193 @oliviamiller could one of you give this a look since rand is out of office?

File: components/board/server.go [hunk 2]
```diff
@@ -218,6 +261,7 @@ func (s *serviceServer) StreamTicks(
 		if err != nil {
 			return err
 		}
+
```
Comment (JohnN193) [inline components/board/server.go:264]:
```suggestion
```

File: components/board/server.go [hunk 3]
```diff
@@ -11,6 +12,21 @@ import (
...
+	}
+	// ErrAnalogByNameReturnNil is the error returned when an analog is nil.
+	ErrAnalogByNameReturnNil = func(boardName string) error {
+		return fmt.Errorf("board component %v AnalogByName should not return nil analog", boardName)
+	}
+	// ErrDigitalInterruptByNameReturnNil is the error returned when a digital interrupt is nil.
+	ErrDigitalInterruptByNameReturnNil = func(boardName string) error {
+		return fmt.Errorf("board component %v DigitalInterruptByName should not return nil digital interrupt", boardName)
+	}
+)
```
Comment (JohnN193) [inline components/board/server.go:28]: only real thought I'm having is that these could live in a separate errors.go file within the board interface. Def not necessary though.

**Review by JohnN193 — APPROVED** [review]
LGTM, though I am still alittle worried about the client not needing nil checks still, but since we use modules for everything I don't think it should be a problem. 

for modules everything running through a viam-server effectively goes through the server.go code here, so that should make sure we don't hit any nil checks in client code

**Comment by martha-johnston** [comment]
> LGTM, though I am still alittle worried about the client not needing nil checks still, but since we use modules for everything I don't think it should be a problem.
> 
> for modules everything running through a viam-server effectively goes through the server.go code here, so that should make sure we don't hit any nil checks in client code

discussed this with Cheuk, actually the modules go first through the client code, but if we return nil from the helper functions, like `GeometriesFromProto` that then goes to the server code and the error is surfaced. for the board ones I'm taking another look now because I agree we might need to just add the checks to client code directly

**Comment by martha-johnston** [comment]
okay just confirmed, we don't need any of the board client checks because the issue that could happen is if a user has a module that returns nil from `GPIOPinByName`, `AnalogByName`, or `DigitalInterruptByName`. The client code has all those functions, but doesn't call them within the other functions that actually set and read pins like the server code does.
